### PR TITLE
Merge feature-branch into development branch

### PR DESCRIPTION
**Pull request by o-rashad**
built an application named Private Events and that behaves similar to Eventbrite events management web application

### CODE REVIEW

**Review by BrianSammit — CHANGES_REQUESTED** [review]
CHANGES REQUIRED :red_circle: :stop_sign: 
--------------------
Hello @od-c0d3r @tanzila-abedin great job you did on this project :muscle: 
TO HIGHLIGHT: 

- The repo is professional
- The project is professional 
- Well structured files 
- Well use of the MVP 

However, there are a few changes to make to complete this project :hammer_and_wrench: see them below :eyes: :point_down: 
----------------

- [x] Check the commit messages. Remember that you must write them with the [formating](https://github.com/trein/dev-best-practices/wiki/Git-Commit-Best-Practices#formatting-rules) rules. You don't need to make the changes this time but keep these formating rules in mind for the next commits. 
- [x] I cannot sign out. Please add the link at the top to be able to sign out. 
- [x] I only can see links to `All Events` and `My events` there is no link to create a new event.
- [x] I cannot create a new event. 

Screenshot from 2021-06-10 17-40-23

- [x] In your event controller in index action you need to list the event in past events, E.g `@past_events = Event.previous_events` and ucomping events, E.g `@upcoming_events = Event.upcoming_events`

past_upcoming_events_requirement

- [x] On the user's show page that you called My events. displays events the user has created and events that the user is attending to separated in past and upcoming.

- [x] In your user's controller in show action you need to list events previoulsy attendented by user using event model scope, E.g `@past_events = current_user.attended_events.previous` and upcoming events previously attended by user using event model scope, E.g `@upcoming_events = current_user.attended_events.upcoming`

user_events_requirement (1)

- [x] On My events I can attend a past event.  

-------------
 Make these required changes and send it again for another review :muscle: :100: 
Reach me on SLACK if you have any questions
@ Brian Sammit Cruz Rodriguez

HAPPY CODING !!!!

AngelicConcreteHypsilophodon-max-1mb

**Comment by o-rashad** [comment]
Thanks @BrianSammit for the review ⭐️⭐️⭐️
We @tanzila-abedin gonna work on 
the required changes and push ASAP.

**Review by BrianSammit — CHANGES_REQUESTED** [review]
CHANGES REQUIRED :red_circle: 
--------
Hello there again :wave: Great job working on the required changes :+1: :100: 
However, there is a minor change to make :wrench: 
--------------

- [ ] There is a bug that is not letting me create a new event. I need to change the controller to make it work but once I change the user the error continues. 

Screenshot from 2021-06-11 11-06-37

Screenshot from 2021-06-11 11-07-50

----------------
Fix this bug on your app and send it again for another review :100: :+1: 
HAPPY CODING !!!

tenor (1)

**Comment by o-rashad** [comment]
@BrianSammit Thanks again ^^, 
We will reach out to you, cuz the bug
is not occurring on my end.

tenor (1)

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  before_action :authenticate_user!, except: %i[index]
+
+  def index
+    @events = Event.all
+    @future_events = @events.upcoming
+    @past_events = @events.past
+  end
+
+  def create
+    @event = current_user.events.create(user_params)
```
Comment (enaburekhan) [inline app/controllers/events_controller.rb]: - [x] uses #build method on association, i.e. `current_user.events.build(event_params)` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/build_method_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)).
Kindly edit the highlighted line of code to reflect :point_up:

File: app/controllers/events_controller.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
+class EventsController < ApplicationController
+  before_action :authenticate_user!, except: %i[index]
+
+  def index
+    @events = Event.all
+    @future_events = @events.upcoming
+    @past_events = @events.past
```
Comment (enaburekhan) [inline app/controllers/events_controller.rb]: - [x] Refactor this code: Currently, the implementation seems difficult to understand. I believe it should be similar to this:
@future_events = Event.upcoming
@past_events = Event.past

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Changes required:wrench::hammer::full_moon:
**Hi team!
Great job so far:clap::clap:. we are almost there. We just need to make a few changes to get the project approved:muscle:.**

**To improve**

- [x] App behavior

- I succeeded in creating just one event. And I couldn't create subsequent events. whether `past` or `upcoming`. I keep getting this [error](https://imgur.com/a/eyWA6t3). Re-check the codes, especially the codes in the controller and fix this bug.

- [x] Controller

- Check my inline comments below and make the required changes.

- [x] Database

- an index for `event_id` column is added to the `tokens` table
- an index for `user_id` column is also added to the `tokens`  table

Why indices for foreign keys are necessary? In short: Since the foreign key columns will be used often to retrieve related information, it's important to index them (queries on those columns will be faster with indexes). More details: check this [article](
https://semaphoreci.com/blog/2017/05/09/faster-rails-is-your-database-properly-indexed.html) and this [article](https://sqlperformance.com/2012/11/t-sql-queries/benefits-indexing-foreign-keys).

Make the above-required changes and when you are ready, request another review.

Do have questions/comments? kindly drop them :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy hacking:blush:
Eric Enaburekhan.

**Comment by o-rashad** [comment]
Thank you @enaburekhan for the review ^^!
We worked on the changes.
The app works fine now.

File: app/controllers/events_controller.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
+class EventsController < ApplicationController
+  before_action :authenticate_user!, except: %i[index]
+
+  def index
+    @future_events = Event.all.upcoming
+    @past_events = Event.all.past
```
Comment (rindrajosia) [inline app/controllers/events_controller.rb]: Refactor this code: Currently, the implementation seems difficult to understand. I believe it should be similar to this:
@future_events = Event.upcoming
@past_events = Event.past

File: README.md [hunk 4]
```diff
@@ -1,24 +1,67 @@
...
-* Deployment instructions
+Then install the needed packages (while skipping any Ruby gems needed only in production):
 
-* ...
+```
+$ bundle install
+```
+
+Next, migrate the database:
+
```
Comment (rindrajosia) [inline README.md:29]: [OPTIONAL] Please ask user to create db via `rails db:create `before migration

File: README.md [hunk 5]
```diff
@@ -1,24 +1,67 @@
...
+```
+$ rails db:migrate
+```
```
Comment (rindrajosia) [inline README.md:32]: [OPTIONAL] You can ask the user to use command `rails s `after migration to open your app.

**Review by rindrajosia — CHANGES_REQUESTED** [review]
Changes Required :red_circle:
To Highlight: :angel:

- Logged in user can create a new upcoming event  :+1: 

To improve :factory_worker: [REQUIRED]:

- [x] In your event controller in index action you need to list the event in past events, E.g @past_events = Event.upcoming and ucomping events, E.g @upcoming_events = Event.past ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))(Please check my inline comments below :arrow_double_down: )
- [x] an index for `creator_id, attendee_id and event_id` columns should added. Since the foreign key columns will be used often to retrieve related information, it's important to index them (queries on those columns will be faster with indexes).

- [x] I cannot create Past events. Please fix it.



- [x] Please, remove **Attend link**  when user is not `logged in `because only `logged in` user can attend events.



Nice to have:

- [x] routes only for used actions, `i.e. resources :events, only: [:new, :create, :show, :index]`

That's it, you can submit another code review request after making these changes :recycle: :innocent:
With all my respect :pray:
Rindra Josia

**Review by oloomoses — APPROVED** [review]
# Congratulations, Project Approved. :muscle: :tada: :clap: 
**Hello team,
Great job working on the suggested changes from the previous review.
You have passed all the requirements for this project :muscle: 
Feel free to merge and close this pull request.**

**I wish you the best in your next project and see you at the top :rocket:** 
#### Happy Coding
_Slack: Oloo Moses_

**Pull request merged**